File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* root , int& h)
    {
        if(root==NULL)
           return 0;
        int lefty = 0, righty=0;
        lefty = height(root->left,h);
        righty = height(root->right,h);
        h = max(h,lefty+righty);
        return 1+max(lefty,righty);   
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int maxi = 0;
        height(root,maxi);
        return maxi;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def helper(root):
            if not root:
                return 0,0
            lft , lfth = helper(root.left)
            rgt , rgth = helper(root.right)

            dia = max(lfth +rgth, rgt, lft) 
            return dia , max(lfth , rgth)+1
        ans , _ = helper(root)
        return ans          





        